phoenix-core/src/main/java/org/apache/phoenix/schema/transform/Transform.java

```diff
@@ -379,71 +421,265 @@ public static void upsertTransform(
+                    pOldTable.getEncodingScheme() != pNewTable.getEncodingScheme()) {
+                MetaDataClient.mutateTransformProperties(connection, tenantId, schema, tableName, newTableName,
+                        pNewTable.getImmutableStorageScheme(), pNewTable.getEncodingScheme());
+                // We need to update the columns's qualifiers as well
+                mutateColumns(connection.unwrap(PhoenixConnection.class), pOldTable, pNewTable);
+
+                // Also update view column qualifiers
+                for (TableInfo view : childViewsResult.getLinks()) {
+                    PTable pView = PhoenixRuntime.getTable(connection, view.getTenantId()==null? null: Bytes.toString(view.getTenantId())
+                            , SchemaUtil.getTableName(view.getSchemaName(), view.getTableName()));
+                    mutateViewColumns(connection.unwrap(PhoenixConnection.class), pView, pNewTable);
+                }
+            }
+            connection.commit();
+
+            // We can have millions of views. We need to send it in batches
+            int maxBatchSize = connection.getQueryServices().getConfiguration().getInt(MUTATE_BATCH_SIZE_ATTRIB, QueryServicesOptions.DEFAULT_MUTATE_BATCH_SIZE);
+            int batchSize = 0;
+            for (TableInfo view : childViewsResult.getLinks()) {
+                String changeView = String.format(changeViewStmt,
+                        (columnNames.size() > 0? "," +String.join(",", columnNames):""),
+                        (view.getTenantId()==null || view.getTenantId().length == 0? null: ("'" + Bytes.toString(view.getTenantId()) + "'")),
+                        (view.getSchemaName()==null || view.getSchemaName().length == 0? null : ("'" + Bytes.toString(view.getSchemaName()) + "'")),
+                        Bytes.toString(view.getTableName()),
+                        (columnValues.size() > 0? "," +String.join(",", columnValues):""));
+                LOGGER.info("Cutover changing view via " + changeView);
+                connection.createStatement().execute(changeView);
+                viewsToUpdateCache.add(view);
+                batchSize++;
+                if (batchSize >= maxBatchSize) {
+                    connection.commit();
+                    batchSize = 0;
+                }
+            }
+            if (batchSize > 0) {
+                connection.commit();
+                batchSize = 0;
+            }
+
+            connection.unwrap(PhoenixConnection.class).getQueryServices().clearCache();
+            UpgradeUtil.clearCacheAndGetNewTable(connection.unwrap(PhoenixConnection.class),
+                    connection.getTenantId()==null?null:connection.getTenantId().getString(),
+                    schema, tableName, systemTransformRecord.getLogicalParentName(), MIN_TABLE_TIMESTAMP);
+            for (TableInfo view : viewsToUpdateCache) {
+                UpgradeUtil.clearCacheAndGetNewTable(connection.unwrap(PhoenixConnection.class),
```

Comment:
Isn't this potentially a million calls to system catalog? Is there a way we can bulk expire the cache if the views are too numerous? Or just expire without getting the new view PTable until the next time someone needs to query it?

Reply:
Good idea. Let me just expire them without querying.